### PR TITLE
Sof hda series for legacy hda mode enabling.

### PR DESCRIPTION
**Pull request by keyonjie**
it's still quite drafted, but with it, legacy mode playback goes(with no sound).

### CODE REVIEW

File: sound/soc/sof/core.c [hunk 1]
```diff
@@ -291,6 +291,8 @@ static int sof_probe(struct platform_device *pdev)
 		goto ipc_err;
 	}
 
+/* bypass DSP if in force legacy hda debug mode */
+#ifndef CONFIG_SND_SOC_SOF_FORCE_LEGACY_HDA
```
Comment (plbossart) [inline sound/soc/sof/core.c:295]: or DSP is not enabled, you could test this on SKL!

File: sound/soc/sof/core.c [hunk 2]
```diff
@@ -324,13 +327,16 @@ static int sof_probe(struct platform_device *pdev)
 		goto comp_err;
 	}
 
+/* bypass DSP if in force legacy hda debug mode */
+#ifndef CONFIG_SND_SOC_SOF_FORCE_LEGACY_HDA
```
Comment (plbossart) [inline sound/soc/sof/core.c:331]: make this a variable that is enabled if IS_ENABLED(FORCE_LEGACY_HDA) || (DSP not enabled)

File: sound/soc/sof/core.c [hunk 3]
```diff
@@ -346,9 +352,12 @@ static int sof_probe(struct platform_device *pdev)
...
 	snd_sof_free_topology(sdev);
 fw_run_err:
+/* bypass DSP if in force legacy hda debug mode */
+#ifndef CONFIG_SND_SOC_SOF_FORCE_LEGACY_HDA
 	snd_sof_fw_unload(sdev);
 fw_load_err:
 	snd_sof_ipc_free(sdev);
+#endif
 ipc_err:
 	snd_sof_free_debug(sdev);
```
Comment (plbossart) [inline sound/soc/sof/core.c:362]: isn't the debug tied to DSP?

File: sound/soc/sof/intel/Kconfig [hunk 4]
```diff
@@ -113,4 +113,15 @@ config SND_SOC_SOF_HDA
...
 
+config SND_SOC_SOF_FORCE_LEGACY_HDA
+	bool "SOF force legacy hda and bypass DSP"
+	depends on SND_SOC_SOF_HDA
+	depends on SND_SOC_SOF_DEBUG
+	help
+          This forces the driver to use HDA legacy mode, and bypasses
+          the embedded DSP in HD-A controller on Intel SKL+ platforms.
+		  This is designed for debugging only.
+          Say Y if you do wan't to use this mode.
```
Comment (plbossart) [inline sound/soc/sof/intel/Kconfig:124]: too many double negations or typos?

File: sound/soc/sof/intel/hda-pcm.c [hunk 5]
```diff
@@ -136,6 +137,38 @@ int hda_dsp_pcm_trigger(struct snd_sof_dev *sdev,
...
+	struct hdac_bus *bus = sof_to_bus(sdev);
+	snd_pcm_uframes_t pos = 0;
+
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
+		pos = snd_sof_dsp_read(sdev, HDA_DSP_HDA_BAR,
+			AZX_REG_VS_SDXDPIB_XBASE +
+				(AZX_REG_VS_SDXDPIB_XINTERVAL *
+				hstream->index));
+	} else {
+		udelay(20);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-pcm.c:154]: add comment here on the difference between capture and playback.

File: sound/soc/sof/intel/apl.c [hunk 6]
```diff
@@ -81,6 +81,7 @@ struct snd_sof_dsp_ops sof_apl_ops = {
 	.pcm_close	= hda_dsp_pcm_close,
 	.pcm_hw_params	= hda_dsp_pcm_hw_params,
 	.pcm_trigger	= hda_dsp_pcm_trigger,
+	.pcm_pointer	= hda_dsp_pcm_pointer,
```
Comment (plbossart) [inline sound/soc/sof/intel/apl.c:84]: Aren't these two patches completely unrelated to legacy HDA?

File: sound/soc/sof/ops.h [hunk 7]
```diff
@@ -265,6 +266,17 @@ snd_sof_pcm_platform_trigger(struct snd_sof_dev *sdev,
 		return 0;
 }
 
+/* host stream pointer */
```
Comment (plbossart) [inline sound/soc/sof/ops.h:269]: same, isn't this something we should have regardless of legacy HDA or not?

File: sound/soc/sof/pcm.c [hunk 8]
```diff
@@ -338,8 +339,10 @@ static snd_pcm_uframes_t sof_pcm_pointer(struct snd_pcm_substream *substream)
 
 	dev_dbg(sdev->dev, "PCM: stream %d dir %d DMA position %lu DAI position %lu\n",
 		spcm->pcm.pcm_id, substream->stream, host, dai);
-
 	return host;
+#else
+	return	snd_sof_pcm_platform_pointer(sdev, substream);
+#endif
```
Comment (plbossart) [inline sound/soc/sof/pcm.c]: Ah so this one is very interesting. Why not rely on DPIB all the time? Is there a reason why we need to reason from the DSP?
It really looks like either the DSP provides the information, or the DMA does it through the pointer information. This should be a platform property, not hardcoded with a define.

File: sound/soc/sof/intel/hda-stream.c [hunk 9]
```diff
@@ -322,8 +322,12 @@ int hda_dsp_stream_hw_params(struct snd_sof_dev *sdev,
 	/* decouple host and link DMA */
 	mask = 0x1 << hstream->index;
 	snd_sof_dsp_update_bits(sdev, HDA_DSP_PP_BAR, SOF_HDA_REG_PP_PPCTL,
+#ifndef CONFIG_SND_SOC_SOF_FORCE_LEGACY_HDA
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-stream.c:325]: again use a variable (or a field in the device structure).

File: sound/soc/sof/intel/hda-stream.c [hunk 10]
```diff
@@ -526,6 +530,7 @@ irqreturn_t hda_dsp_stream_threaded_handler(int irq, void *context)
 			    !s->running ||
 			    (sd_status & SOF_HDA_CL_DMA_SD_INT_MASK) == 0)
 				continue;
+			snd_pcm_period_elapsed(s->substream);
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-stream.c]: Ah very interesting. We had a discussion that there should be only one source for period elapsed, but here you are adding another one? Would this work in normal mode w/ the DSP enabled?

File: sound/soc/sof/ipc.c [hunk 11]
```diff
@@ -264,6 +264,10 @@ int sof_ipc_tx_message(struct snd_sof_ipc *ipc, u32 header,
 	struct snd_sof_ipc_msg *msg;
 	unsigned long flags;
 
+	/* todo: fix me: temporary disable ipc message sending */
```
Comment (plbossart) [inline sound/soc/sof/ipc.c:267]: why is this needed? You didn't initialize IPC in the first patches so would this code even be used?

File: sound/soc/intel/boards/bxt_tdf8532.c [hunk 12]
```diff
@@ -10,6 +10,19 @@
 #include <sound/pcm.h>
 #include <sound/soc.h>
 #include <sound/pcm_params.h>
+#include <sound/jack.h>
```
Comment (plbossart) [inline sound/soc/intel/boards/bxt_tdf8532.c:13]: If the DSP is not enabled, we should select another machine driver.

File: sound/soc/sof/intel/Kconfig (hunk 4 above)
Comment (keyonjie) [reply sound/soc/sof/intel/Kconfig:124]: it's typo here, should be "Say Y if you do want to use this mode."

File: sound/soc/sof/core.c (hunk 1 above)
Comment (keyonjie) [reply sound/soc/sof/core.c:295]: OK, I will add checking of DSP's availability.

File: sound/soc/sof/core.c (hunk 2 above)
Comment (keyonjie) [reply sound/soc/sof/core.c:331]: OK, will change it.

File: sound/soc/sof/core.c (hunk 3 above)
Comment (keyonjie) [reply sound/soc/sof/core.c:362]: I need refine this part, and adding check to dsp availability also.

File: sound/soc/sof/intel/hda-pcm.c (hunk 5 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-pcm.c:154]: OK, will do.

File: sound/soc/sof/pcm.c (hunk 8 above)
Comment (keyonjie) [reply sound/soc/sof/pcm.c]: that's true, the new update is that the DPIB works for DSP I2S also on my local platform, I will run more test and if no regression then we can switch to use DPIB for all mode(I still want to keep IPC position update as another choice though).

File: sound/soc/sof/intel/hda-stream.c (hunk 9 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c:325]: you mean the same as above with adding dsp availability check?

File: sound/soc/sof/intel/hda-stream.c (hunk 10 above)
Comment (keyonjie) [reply sound/soc/sof/intel/hda-stream.c]: to use DPIB mode, we need add this and remove the IPC one. Looks the removal of IPC one missed? let me update it.

File: sound/soc/sof/ipc.c (hunk 11 above)
Comment (keyonjie) [reply sound/soc/sof/ipc.c:267]: the topology parsing still try to send IPC even neither of dsp and IPC is initialized.
this is hack, let me refine it.

File: sound/soc/intel/boards/bxt_tdf8532.c (hunk 12 above)
Comment (keyonjie) [reply sound/soc/intel/boards/bxt_tdf8532.c:13]: that's true, will change like that.

**Comment by keyonjie** [comment]
this is outdated, we have already another PR been merged, legacy HDA mode is not ready yet. let me close this one.
